Implementing logic to identify primary host

fabiocav opened the pull request:
Resolves #526 
This is a requirement for #164 and other items.

fabiocav commented:
@brettsam if you have a chance, can you take a look at this?

brettsam commented on test/WebJobs.Script.Tests/BlobLeaseManagerTests.cs:
```diff
@@ -0,0 +1,209 @@
+        [Fact]
+        [Trait("Category", "E2E")]
+        public async Task HasLeaseChanged_WhenLeaseStateChanges_IsFired()
+        {
```
It'd be good to have a test the other way as well -- make sure it fires when you lose a lease.

fabiocav replied:
Good point. I'll add a test for that scenario.

brettsam commented on src/WebJobs.Script/Host/BlobLeaseManager.cs:
```diff
@@ -0,0 +1,173 @@
+        }
+
+        private async Task AcquireOrRenewLeaseAsync()
+        {
```
What happens if the Contaner or Blob disappears out from under us? Should you recreate what we need if that happens?

fabiocav replied:
Yes, we should deal with that scenario. I'll make that change.

brettsam commented on src/WebJobs.Script/Host/BlobLeaseManager.cs:
```diff
@@ -0,0 +1,173 @@
+                    _timer.Change(_renewalInterval, _renewalInterval);
+                }
+            }
+            catch (StorageException exc) when (exc.RequestInformation.HttpStatusCode == 409)
```
>catch (StorageException exc) when (exc.RequestInformation.HttpStatusCode == 409) [](start = 12, length = 80)

Should we remove the LeaseId if any exception is thrown here? Or is that handled some other way?

Example -- the primary instance can't connect for some reason and times out trying to renew. It will still have a LeaseId even though it'll time out in a few seconds. There may be the possibility of two hosts with HasLease = true for a short time.

fabiocav replied:
We're letting any other exceptions bubble up here, but I'll make some changes to make this more resilient to transient changes and also always reset the flag.

fabiocav commented on src/WebJobs.Script/Host/BlobLeaseManager.cs:
```diff
@@ -0,0 +1,173 @@
+            if (!await blob.ExistsAsync())
+            {
+                await blob.UploadFromStreamAsync(new MemoryStream());
+            }
```
I'm assuming you're talking about line 132 here. Yes, there is a chance. I'll address this.

fabiocav commented:
@brettsam addressed feedback, made some additional enhancements and added a couple of extra tests. Let me know if you notice anything else we may need to change.

fabiocav commented on src/WebJobs.Script/Host/ScriptHost.cs:
```diff
@@ -89,6 +93,12 @@ public AutoResetEvent RestartEvent
+            // this is for testing only
+            _standbyMode = null;
+        }
+
```
Moved this because StyleCop was complaining about the order (properties before methods)

brettsam commented on src/WebJobs.Script/Host/ScriptHost.cs:
```diff
@@ -45,6 +46,8 @@ protected ScriptHost(ScriptHostConfiguration scriptConfig)
+        public event EventHandler IsPrimaryChanged;
+
         public static bool InStandbyMode
         {
```
Can you explain StandbyMode? Maybe it needs a comment?

fabiocav replied:
This was part of the hot standby change made here https://github.com/Azure/azure-webjobs-sdk-script/pull/534
It's a flag that indicates whether the host is in standby mode, which means it's pre-warmed but not setup with an actual customer Function App settings/environment.

fabiocav replied:
I'll document it outside of this PR.

brettsam commented on test/WebJobs.Script.Tests/BlobLeaseManagerTests.cs:
```diff
@@ -0,0 +1,316 @@
+            
+            var blobMock = new Mock<ICloudBlob>();
+            blobMock.Setup(b => b.AcquireLeaseAsync(It.IsAny<TimeSpan>(), It.IsAny<string>()))
+                .Returns(() => results.Dequeue());
```
Moq has a 'SetupSequence' method for stuff like this. No need to change this, just thought I'd point it out in case you hadn't seen it before.

fabiocav replied:
I did go down the path of using `SetupSequence` (and would prefer to use that) but changed it because it lacks a convenient mechanism to enable the logic we're performing in the callback to signal the `ManualResetEventSlim`. I might have missed something, though, so I'm open to ideas.

brettsam commented:
:shipit:

fabiocav commented:
@brettsam merged, but left responses to your last comments. If needed, I'll make the changes to the test.
